Document <LatitudeEmbed /> (#286)
latitude-dev/latitude: docs/guides/embed/react-embed.mdx

description: 'Embedding a Latitude data app within your application'
---

## Introduction

When developing a web application with React, you have the option to embed a Latitude data app via a dedicated React component. This method allows for a more interactive embedding experience and provides the capability to manage the app using JavaScript.

There are several benefits to this approach:

<Note>
  1. Enables your React app to notify the Latitude app when a parameter is modified within your React app.
  2. Allows for the reverse, in which the Latitude app communicates parameter changes back to your React app.
  3. Gives you the ability to execute all or selected queries directly from your React app.
  4. Permits the dispatch of custom events from your React app to the Latitude view.
</Note>

We have an [example application](https://github.com/latitude-dev/latitude/tree/main/examples/sample-react). You can check the source code if you get lost in the process.
In [Embedding/index.tsx](https://github.com/latitude-dev/latitude/blob/main/examples/sample-react/src/examples/Embedding/index.tsx) you can see all the options you can pass to the `<LatitudeEmbed />` component.

> Try to change in this example the field in `YOUR REACT APP` and notice how the Latitude app changes. Now do the same but inside `LATITUDE IFRAME` and notice how the React app changes. Click `Run` button, this will run all the queries in the example.

Reviewer: This part is not super clear

Author: I changed it anyway users will have the context of the example app
![image](https://github.com/latitude-dev/latitude/assets/49499/dbff756d-a53a-4699-b6d8-f7b29c705489)

And lastly `Pick your character` and see how Latitude app shows your Star Wars character. May the force be with you!

<iframe class="rounded-xl" src="https://latitude-embed.vercel.app/embedding" width="100%" height="800"></iframe>

But let's go through the options you can pass to the `<LatitudeEmbed />` component.

> Let's walk through the process of embedding a Latitude data app within a React application using the `<LatitudeEmbed />` component.

Reviewer: same thing twice

## Installation
You need to install our React SDK

```bash
npm install @latitude-app/react
```

In the root of your app you need to initialize the embedding component with `defineCustomElements(window)` method.
This is necessary because under the hood we are using web components and they need to be initialized before using them.

```tsx src/index.tsx
```tsx src/app.tsx
import { defineCustomElements } from '@latitude-data/react'
import App from './App'

// This line is necessary
defineCustomElements(window)

ReactDOM.createRoot(document.getElementById('root')!).render(
  <React.StrictMode>
    {/* Your app goes here */}
    <App />
  </React.StrictMode>,
)
```

> And lastly use `<LatitudeEmbed />` component to embed the Latitude app.

Reviewer: ```suggestion
And lastly use the `<LatitudeEmbed />` component to embed the Latitude app.
```

  const [endYear, setEndYear] = useState('2005')

  return (
    <>
      <input
        type='text'
        value={endYear}
        onChange={(e) => {
          // Use this method to change the parameters of the Latitude app
          changeEmbedParams({ end_year: value })
          setEndYear(e.target.value)
        }}
      />
      <LatitudeEmbed
        url='http://localhost:3000/some/view'
        params={{ start_year: '2003', end_year: endYear }}
      />
    </>
  )
}
```

### Listen to parameter changes
You can listen to parameter changes in the Latitude app and update your React app accordingly.

```tsx src/App.tsx
export default function App () {
  const [endYear, setEndYear] = useState('2005')

  return (
    <LatitudeEmbed
      url='http://localhost:3000/some/view'
      params={{ start_year: '2003', end_year: endYear }}
      onParamsChanged={(event) => {
        const params = event.detail.params
        const newEndYear = params.end_year as string
        if (endYear === newEndYear) return

        // Update the end year in your React app
        // when the end year in the Latitude app changes
        setEndYear(newEndYear)
      }}
    />
  )
}
```

> ### Execute queries

Reviewer: ```suggestion
### Run queries
```

> You can execute all or selected queries directly from your React app.

Reviewer: ```suggestion
You can run all queries or specific ones directly from your React app.
```
